### ISSUE TITLE
Unexpected behaviour of move_plans_between_datasets and confusing documentation

### ISSUE THREAD

**Issue by sarahannblackledge** [MIC-DKFZ/nnUNet#1970]
Hello,

First of all, thank you so much for releasing nnUNet. It is a very amazing tool!

I have a question regarding the documentation and usage of the move_plans_between_dataset function for the purpose of pre-training and fine-tuning. 

1. Is the mismatch between source and target flags and user inputs intentional in the following? This is the code snippet copied from your documentation on pre-training and fine-tuning.

`nnUNetv2_move_plans_between_datasets -s TARGET_DATASET -t SOURCE_DATASET -sp TARGET_PLANS_IDENTIFIER -tp SOURCE_PLANS_IDENTIFIER`

2. Regardless of how I run the code (the way indicated above, or the intuitive way shown below), I am unable to run preprocessing on the source task because there is no plan named with the TARGET_PLANS_IDENTIFIER in the SOURCE_DATASET. 

`nnUNetv2_move_plans_between_datasets -t TARGET_DATASET -s SOURCE_DATASET -tp TARGET_PLANS_IDENTIFIER -sp SOURCE_PLANS_IDENTIFIER`

Can you please elaborate on the expected behaviour? I've had a look at the move_plans_between_datasets.py file, and can see that it effectively overwrites any dictionary key referencing the -s dataset with the -t dataset in the -s nnUNetPlans.json file. The confusing thing for me is that it is saved in the -t directory, rather than the -s directory, so naturally it's not found when running the following:

`nnUNetv2_preprocess -d SOURCE_DATSET -plans_name TARGET_PLANS_IDENTIFIER`

It would be really helpful if you could explain what's meant to happen, and what files should be saved where. 

Thanks!

Sarah

**Comment by constantinulrich** [comment]
Hi :) i was struggeling myself several times about the docu. 
 checkout this related [issue](https://github.com/MIC-DKFZ/nnUNet/issues/1477): 

> The reason for that is the double meaning of source and target in this context. We have source (pretraining) and target (finetuning) datasets, but we also have the function nnUNetv2_move_plans_between_datasets where you move plans from one dataset (arg source) to the other (arg target). Since you want to move the plans from the dataset you want to finetune on (target) to the pretraining dataset (source) you end up with -s TARGET_DATASET -t SOURCE_DATASET

**Comment by constantinulrich** [comment]
>  I've had a look at the move_plans_between_datasets.py file, and can see that it effectively overwrites any dictionary key referencing the -s dataset with the -t dataset in the -s nnUNetPlans.json file. The confusing thing for me is that it is saved in the -t directory, rather than the -s directory

The idea of this function is to transfer the plans from source dataset to target dataset. (not to be confused with source and target datasets when you are doing pretraining). Therefore, the plan should be saved in the -t folder, but containing the network configuration, normalization and target spaving of the -s plan. 

Best,
Constantin

**Comment by sarahannblackledge** [comment]
Hi Constantine,

Thank you so much for the quick response. I see now how Source and Target have double meanings.

I'm still struggling though - apologies if I'm being very thick!

You wrote the following in your previous comment:
_The idea of this function is to transfer the plans from source dataset to target dataset. (not to be confused with source and target datasets when you are doing pretraining). Therefore, the plan should be saved in the -t folder, but containing the network configuration, normalization and target spacing of the -s plan._

I indeed see the plan in my -t folder, and it appears that everything with the move_plans_between_datasets function is working correctly as you've described. The problem arises when I come to the next step of running the preprocessing on the source task:

`nnUNetv2_preprocess -d SOURCE_DATSET -plans_name TARGET_PLANS_IDENTIFIER`

Can you clarify what the SOURCE_DATASET is? I have originally assumed the SOURCE_DATASET was the dataset in the -s folder. However, I get a 'No such file or directory" error message when I run the above nnUNetv2_preprocess command, as it is looking for the plan indicated by TARGET_PLANS_IDENTIFIER in the -s folder. As the plan is saved in the -t folder during the nnUNetv2_move_plans_between_datasets folder, the error makes sense. What doesn't make sense to me is where I am going wrong! Perhaps the SOURCE_DATASET is supposed to refer to the dataset in my -t directory? I'm really sorry, I'm probably getting my wires crossed with the terminology. I'd really appreciate further clarification if possible.

**Comment by constantinulrich** [comment]
Hi, 
no worries, it is really confusing...
> nnUNetv2_preprocess -d SOURCE_DATSET -plans_name TARGET_PLANS_IDENTIFIER

Here, -d SOURCE_DATSET is the one you are planning to do the preprocessing on, which is the target dataset from the move_plans function. 

In general, source_dataset is the pertaining dataset and target dataset is the one you wanna fine-tuning on. 
But in the move_plans function it is the other way round.

**Comment by sarahannblackledge** [comment]
Hi Constantine,

I finally figured out where I was going wrong. In a nutshell, I was swapping around the plan identifiers between source and target. If I understand correctly, I think there might be a few instances in the documentation where TARGET_PLANS_IDENTIFIER should be swapped with SOURCE_PLANS_IDENTIFIER. I'll elaborate below in case this is of interest, but no need to take further action for my sake. Thanks so much for your help and guidance!

I'll specify my dataset names and plan identifiers for the purpose of clarity and for following along with my sample code:

`SOURCE_DATASET=013 #(the dataset I intend to run pretraining on)
TARGET_DATASET=014 #(the dataset I am interested in; the one I wish to fine tune on). 
SOURCE_PLANS_IDENTIFIER=nnUNetPlans_test
TARGET_PLANS_IDENTIFIER=Tx1_test_nnUNetplans`

Step 1: Run experiment planning (and preprocessing) on target dataset. This will create a plan.json file with the name specified by TARGET_PLANS_IDENTIFIER. If unspecified, the plan will be named 'nnUNetPlans' by default.

`nnUNetv2_plan_and_preprocess -d -overwrite_plans_name $TARGET_PLANS_IDENTIFIER`

Step 2: Extract dataset fingerprint of the source dataset. I ran this exactly as the original documentation suggested
`nnUNetv2_extract_fingerprint -d $SOURCE_DATASET`

Step 3: Take the plans from the target dataset and transfer it to the source. Here, it's important to customize SOURCE_PLANS_IDENTIFIER. This is the name of the plan that will be saved in your -t directory (which is where the SOURCE_DATASET is saved in my case).  If you do not change this, the nnUNetPlans.json file that you might have where you originally trained the network will be overwritten. The documentation does warn that this may happen, but recommends changing the TARGET_PLANS_IDENTIFIER, which I think is confusing as this is not the one in the -t directory.

`nnUNetv2_move_plans_between_datasets -s $TARGET_DATASET -t $SOURCE_DATASET -sp $TARGET_PLANS_IDENTIFIER -tp $SOURCE_PLANS_IDENTIFIER`

Up until this point, my code hasn't deviated from the documentation. That part comes now...

Step 4: Run preprocessing on the source task. Here, it will look in the -t directory for a plans json file with the SOURCE_PLANS_IDENTIFIER name. Note, this is different than the documentation, as the documentation suggests that the -plans_name flag should be TARGET_PLANS_IDENTIFIER.

`nnUNetv2_preprocess -d $SOURCE_DATASET -plans_name $SOURCE_PLANS_IDENTIFIER`

Step 5: Run the training as usual. Here, I also had to change the -p flag to be SOURCE_PLANS_IDENTIFIER rather than TARGET_PLANS_IDENTIFIER as suggested by the documentation

Anyways, this is what's working for me so far. I hope my feedback is helpful. Thanks again for the amazing tool. nnUNet is amazing!
`nnUNetv2_train SOURCE_DATSET CONFIG all -p TARGET_PLANS_IDENTIFIER`

**Comment by constantinulrich** [comment]
Hi, thanks for your feedback. I will forward it and we will change the docu soon. Glad that it works now for you!
